File: modules/randh.c
#include <stdlib.h>
#include "utone.h"

int ut_randh_create(ut_randh **p)
{
    *p = malloc(sizeof(ut_randh));
    return UT_OK;
}

int ut_randh_destroy(ut_randh **p)
{
    free(*p);
    return UT_OK;
}

int ut_randh_init(ut_data *ut, ut_randh *p)
{
    p->counter = 0;
    p->freq = 10;
    p->dur = (ut->sr / p->freq);
    p->min = 0;
    p->max = 1;
    p->val = 0;
    return UT_OK;
}

int ut_randh_compute(ut_data *ut, ut_randh *p, UTFLOAT *in, UTFLOAT *out)
{
    if(p->counter == 0) {
        p->val = p->min + ((UTFLOAT) ut_rand(ut) / UT_RANDMAX) * (p->max - p->min);
        
        if(p->freq == 0) {
            p->dur = 1;
        } else {
            p->dur = (ut->sr / p->freq) + 1; 
        }

        *out = p->val;
    } else {
        *out = p->val;
    }
    p->counter = (p->counter + 1) % p->dur;
    return UT_OK;
}

File: modules/data/randh.lua
sptbl["randh"] = {

    files = {
        module = "randh.c",
        header = "randh.h",
        example = "ex_randh.c",
    },

    func = {
        create = "ut_randh_create",
        destroy = "ut_randh_destroy",
        init = "ut_randh_init",
        compute = "ut_randh_compute",
    },

    params = {
        optional = {
            {
                name = "min",
                type = "UTFLOAT",
                description ="Minimum value to use.",
                default = 0
            },
            {
                name = "max",
                type = "UTFLOAT",
                description ="Maximum value to use.",
                default = 1
            },
            {
                name = "freq",
                type = "UTFLOAT",
                description = "Frequency of randomization (in Hz)",
                default = 10
            },
        }
    },

    modtype = "module",

    description = [[Random number generator with hold time.
Randh is loosely based off of the Csound opcode randomh. The design is equivalent
to scaled noise sent through a classic sample and hold module.
]],

    ninputs = 0,
    noutputs = 1,

    inputs = {
    },

    outputs = {
        {
            name = "out",
            description = "Signal output."
        },
    }

}
